include/seastar/coroutine/exception.hh

```diff
@@ -104,10 +117,19 @@ exception make_exception(T&& t) {
 /// ```
 /// co_await coroutine::return_exception(std::runtime_error("something failed miserably"));
 /// ```
-template<typename T>
 [[nodiscard]]
-exception return_exception(T&& t) {
-    return exception(std::make_exception_ptr(std::forward<T>(t)));
+inline exception return_exception(std::exception_ptr&& ex) noexcept {
+    return exception(std::move(ex));
+}
+
+[[nodiscard]]
+inline exception return_exception(const std::exception_ptr& ex) noexcept {
+    return exception(ex);
+}
+
+[[nodiscard]]
+exception return_exception(std::derived_from<std::exception> auto&& e) noexcept {
+    return exception(std::move(e));
```

Comment:
> Isn't std::forward actually needed because this is a template, not an rvalue reference?

Yes, taking the template <typename E> out, we cannot std::forward<E>(e), just std::move it.
I don't think this makes any difference in performance since e is passed be rvalue reference, not by value, so its move constructor or assignment operator are not called until it lands at its destination.

Reply:
Isn't "auto&&" a template?
In other words, isn't this && a sign of a **universal reference**, _not_ an **rvalue reference**?
A quick Google search yielded https://stackoverflow.com/questions/13230480/what-does-auto-tell-us which suggests maybe it's not an rvalue.

Reply:
Github took the liberty to hide stuff in angular brackets in my comment above, so pasting it below with hopefully correct formatting:

> Yes, taking the template `<typename E>` out, we cannot `std::forward<E>(e)`, just std::move it.

so, I don't think we can use [std::forward](https://en.cppreference.com/w/cpp/utility/forward) here.

Reply:
But what I tried to reply that "auto&&" is just another syntax for a template. It's a new (almost) C++20 syntax called "abbreviated function template" - see for example https://ggulgulia.medium.com/c-20-concepts-part-4-389a7649f084

Looking above, it seems Avi asked you to change the template to auto&& just to have shorter syntax - I don't think it changes anything.

Reply:
ok, got you.  We can indeed use `std::forward<decltype(e)>` here.

Reply:
Wow, this looks horrible - i.e., it looks like real C++ ;-)